### DISCUSSION TITLE
Secure Boot

### DISCUSSION THREAD

**Discussion by jbbletterman** [ipxe/ipxe#358] [Q&A] [answered]
# Secure Boot
I'm keeping the title short: "Secure Boot". Why? Because I can barely find anything online about how to use iPXE in conjunction with SB. All I could find on the official iPXE site was about this eToken: https://ipxe.org/appnote/etoken.

## Context
We are trying to boot custom images based on the latest Fedora 34 kernels. These bootable images are created using cpio and dracut. We generate a vmlinuz and initrd file through our pipeline and push them directly to our iPXE server. This works on 90% of our systems (some lack hardware support for anything but Windows). When enabling Secure Boot (SB), the same custom images work (strangely even when removing the keys from the firmware)... SB only works when UEFI is enabled and Legacy options are disabled (Compatibility Support Module) and enabling Secure Boot. Bootable USB's with our images will continue to work, but booting the same images over iPXE tells us that the images aren't validated. I later discovered that iPXE also needs to be signed for SB to function, which I assume is part of the problem.

## Shim, iPXE and Fedora
My understanding so far is that Fedora will boot through SB, as long as the kernel isn't tainted, and the Grub2 and Shim packages are installed. The Fedora Shim package contains the signatures to boot Fedora through SB, which could be why USB-booting works fine with UEFI/SB enabled. The flow looks something like this (I'm guessing): 

`shim.efi` ➡️ `ipxe.efi` ➡️ `vmlinuz (initrd)` ➡️ `Fedora:34` 

I'm having trouble building `shim.efi` and I don't know how to get `ipxe.efi` signed because of this. Apparently Microsoft handles the signing of SB validated images, but there has to be a way to use self-signed ones for testing, right? How do I get ahold of a signed Shim bootloader, to verify our iPXE and boot our images? Are there any guides, tutorials, documents, or are there any people here that know more about using iPXE with SB?

## Windows 8 Ready
Our interest in SB stems from our customers, having purchased "Windows 8 Ready" certified systems (we recommend them not to), some of which come with mandatory SB which cannot be disabled. So far, it seems SB works much more fluently with Windows Deployment Services (WDS), which is no surprise since Microsoft handles the signing of SB images. I've also seen more information about SB in combination with legacy PXE, but we avoid using this because the outdated TFTP protocol prevents us from booting anything at all sometimes.

## Shim Demo
I found this demo, but I fear it might be a little outdated: https://github.com/ipxe/shimdemo. It doesn't seem to work for me; when I try building, it tells me that `efi.h` isn't included. I am a bit lost in all this and I need someone to point me in the right direction.

**Comment by oschroeder8** [comment]
Here is an old thread from the deprecated iPXE forum:
[https://forum.ipxe.org/showthread.php?tid=7533]

There used to be a page in the iPXE.org docs section about the frustrations of trying to get iPXE signed and the lack of a clear process with Microsoft. But I think Michael Brown in the end actually figured out how to reliably get binaries signed by Microsoft since he is doing it for WIMBOOT. (I was trying to find the page since I clearly remember reading it, but I cannot find it anymore on the site.)

Thing is that due to the many ways iPXE can be compiled, with different options, etc. I gather that it is not really possible to offer signed binaries of iPXE.

I have been resorting to using an iPXE binary (snponly.efi) that is signed by VMware from their AutoDeploy package. This works great with Dell servers, but mileage on desktop PCs will probably vary. Unfortunately the signed iPXE binary from VMware does not provide all the options I'd like to use, but it works to boot WIndows WIMs and VMware images in SecureBoot mode - I haven't tried Linux.

Maybe it would be possible to resurrect the page on iPXE.org and have it updated with the established process on getting things signed by Microsoft? There are build options to create a secure-boot-ready iPXE binary that can be submitted for a signature.

Some firmware will allow you to upload custom certificates for secure boot, but again, I work on servers and I am not sure how many desktop PCs will have that option.

You will need to have a signed iPXE binary to successfully secure boot.

**Reply by jbbletterman** [reply]
Thanks for your reply!

I had indeed seen that thread before, but I continued looking around, because I am looking to find a way to test SB with self-signed certificates. We need a proof of concept before we start submitting anything to Microsoft. I knew this process was complex and I can see from your response that it hasn't really gotten any easier. I would indeed love to see the iPXE page updated, I feel like Secure Boot definitely deserves a bit more explanation. As I said: there is hardly anything to find on the internet.

I remember building iPXE for our infrastructure, and testing many different builds until I found a working one on our systems' architectures. It makes sense that this is the reason why it's cumbersome to offer signed binaries.

I also read about the VMware binaries before (as I had seen the deprecated thread), but I strayed away from those. My reason for this is, that we use chainloading scripts to point to an Apache web server for our iPXE configurations (like the menu). I'm assuming the VMware `snponly.efi` doesn't have any embedded scripts, which means we'd have to give up these features. 

For now, I just want a working iPXE solution, preferably without sacrificing any of our current features. I would sign our binaries myself if I could, but I have no idea how to use Shim to bootload into iPXE.

**Reply by we-mi** [reply]
Hi @jbbletterman,

I suggest reading the documentation and install instructions from the "rEFInd" project [here](https://www.rodsbooks.com/refind/index.html). It's a great way to get an overview of how the EFI boot process works and also how to boot shim (signed by Microsoft) and chainload another EFI (signed by yourself). You don't really need rEFInd for what you want to accomplish, it's just for a better understanding.

I was struggling with this too, but came across rEFInd a couple of days ago and am now able to boot our own (not officially signed) ipxe EFI binary with secure boot enabled.

The only downside of this is that you need to first disable secure boot, import the cert or hash for your EFI binary, and re-enable secure boot again on every machine you want this to work.
I am currently working on automating this somehow as much as possible over ipxe but this needs some additional time.

If you have struggles to get this working, send me a PM and I will see how I can help, but keep in mind that I'm at the beginning of this whole "secure boot with ipxe"-thing myself :D

I hope this helps

**Reply by NiKiZe** [reply]
maybe wimboot repo and git commits can be helpfull, there is also an [iPXE cab target](https://github.com/ipxe/ipxe/commit/dbea47ce7d61f253b58aa7268b0261922a6b1a22) to build the file needed to sign?

**Reply by JunielKatarn** [reply]
Could you please point to the VMware AutoDeploy package download location?
Not sure which of these options could contain that software.
https://customerconnect.vmware.com/downloads/info/slug/datacenter_cloud_infrastructure/vmware_vsphere/7_0

**Reply by berlin4apk** [reply]
> Could you please point to the VMware AutoDeploy package download location?

I bet it's somewhere inside of the VMware vCenter Server

**Reply by aggieNick02** [reply]
This page (also by the author of rEFInd) was also really helpful in understanding what's going on and how addition of private keys to multiple computers can be streamlined (as much as possible): https://www.rodsbooks.com/efi-bootloaders/controlling-sb.html

**Reply by fuchsia74** [reply]
> Could you please point to the VMware AutoDeploy package download location? Not sure which of these options could contain that software. https://customerconnect.vmware.com/downloads/info/slug/datacenter_cloud_infrastructure/vmware_vsphere/7_0

You can download in VC auto deploy service:
![image](https://user-images.githubusercontent.com/10117683/136691343-dd12ee43-6d55-4453-9c8b-5bda4444eee2.png)
snponly64.efi.officialkey 
works for me.

**Reply by oschroeder8** [reply]
I just want to call out one obstacle that could be encountered when using this signed binary: This binary was signed with VMware's key, which has a very good chance of being included on server hardware, and you would be able to use it with VMs within a vSphere environment. But there is a possibility that VMware keys would not be included in desktop motherboard firmware as that would not generally be considered an install target for ESXi, etc.
I personally use this with Dell and Lenovo server hardware and it works fine for me.

**Comment by oschroeder8** [comment]
I found the page on ipxe.org that I was referring to in my earlier reply:
[UEFI Secure Boot signing using a DigiCert eToken](https://ipxe.org/appnote/etoken)
As you can see in the *Waiting* section, it must have been frustrating to figure out Microsoft's signing process at the time of writing that.
I believe that @mcb30 has worked through the frustration of the process, because he is getting `WIMBOOT` signed quite regularly and in a consistent manner.
If that process could be documented, it would probably make it easier for us to get our own binaries signed by Microsoft.

**Reply by NiKiZe** [reply]
#488 has several good links from the MS standpoint, and the code review is probably one of the larger hurdles in getting it done.

wimboot is fairly minimal and have very clean code compared to iPXE.

**Reply by fuchsia74** [reply]
> #488 has several good links from the MS standpoint, and the code review is probably one of the larger hurdles in getting it done.
> 
> wimboot is fairly minimal and have very clean code compared to iPXE.

I have tested  2Pint Software signed efi bin file,  it has Embedded script can not working with ipxe script url from dhcpd.
this page ：https://techcommunity.microsoft.com/t5/hardware-dev-center/updated-uefi-signing-requirements/ba-p/1062916
![image](https://user-images.githubusercontent.com/10117683/136704252-dfc6022f-3f43-4e65-a52d-c712b635efd7.png)
But you must be a MS partner.

**Reply by NiKiZe** [reply]
"must be a MS partner" where did you read that?

**Reply by fuchsia74** [reply]
efi bin submit website is ：https://partner.microsoft.com/en-us/dashboard/hardware/filesign , required permission.

**Reply by daberkow** [reply]
I work somewhere that is a partner, and I was told if we arent using iPXE in a product that I could point them to on a website (since this is for system provisioning internally) they would not sign for us.

**Reply by JunielKatarn** [reply]
Guess netboot.xyz could help with that?
Being a separate "product" based on iPXE with its own website?

**Comment by braindevices** [comment]
Personally I think iPXE should never be part of secure boot due to its nature: the idea to use iPXE is to make the boot process as flexible as possible. If we introduce secure boot and measured boot signed by others, it defeats the best part of iPXE.  I think the best we can do is sign the iPXE by ourselves with our own private key instead of signed by MS, then we have full control of our boot process.

I am pretty against the shim idea since it simply defeat the entire idea of secure boot, it just bypass the secure boot. However, since it is already like this, why not just use shim to load iPXE as it does for grub? Then the ones just want to by pass the SB can just by pass it with shim as other people do.

**Reply by NiKiZe** [reply]
You need one shim to load iPXE, then you need a shim again to boot other things, unless they are signed with allowed firmware keys. (That might be many shims) not saying this is an issue. Just something to be aware of.

**Reply by stappersg** [reply]
```text
s/You need one shim to load iPXE/You need a signed shim to load iPXE into a Secure Boot environment,
but only when you are convicted to such environment/
```

That bold text, less bold:
The
> You need one shim to load iPXE

should be
> You need a signed shim to load iPXE into a Secure Boot environment,
> but only when you are convicted to such environment.


@braindevices  wrote
> I think the best we can do is sign the iPXE by ourselves with our own private key instead of signed by MS, then we have full control of our boot process.
> 
> I am pretty against the shim idea since it simply defeat the entire idea of secure boot, it just bypass the secure boot.


I think that I don't understand @braindevices or that @braindevices has a solution for different problem that is discussed here. But that is not important. Important is finding / building something that works.

**Reply by stappersg** [reply]
Link to #300. Opening text of that _merge request_.
> TrenchBoot is a framework that allows individuals and projects to build security engines to perform launch integrity actions for their systems. It includes an open source implementation of DRTM (Dynamic Root of Trust for Measurement), which for AMD platforms starts with SKINIT instruction. After that CPU sends a small piece of code (up to 64kB) to TPM for measurements and starts execution of it. In TrenchBoot, this code is Landing Zone.

**Comment by jerrywoo96** [comment]
Ventoy had managed to implement to secure boot through enrolling a key or hash. afaik, the user only needs to register it once in the machine, and the user can use ventoy with secure boot as many times as he/she want w/o registering it again. I'm not sure about the secure boot process using pxe, but if registering a key or hash to allow ipxe to pass secure boot, its still a solution as we can now install Windows 11. Whether it be a perfect implementation or not, its 1 step forward better than having no solution.

**Reply by NiKiZe** [reply]
Yes, adding your own keys to firmware config is a "good" workaround as long as it is your machine and it is managed in a secure way. (mentioning this so that some manufacturers don't get the idea that they should add their own keys and then leave it there)

**Reply by jerrywoo96** [reply]
maybe our 1st step is to allow us a way to register our own keys in our own machines using a usb drive.

**Reply by jerrywoo96** [reply]
Then build an ipxe image with parameters that works with the key that we enrolled into our system.

**Reply by NiKiZe** [reply]
Late reply here, but yes, essentially all EFI firmware has a way of adding your own keys. When I say this I might be biased towards Enterprise grade hardware, but most machines have ways to update DB (keywords: KEK, DB) even Windows Update does update these with new certificates. (as long as there is not vendor bugs preventing it)

**Comment by GeorgiaM-honestly** [comment]
Has there been any resolution to this? My head spins and I also don't understand all the suggestions here. I'm simply wanting to boot the .img off of a usb drive.

**Reply by horvathcsabalaszlo** [reply]
I also need an ISO booting solution to Secureboot for PXE...

**Comment by NiKiZe** [comment, marked as answer]
2.0.0 is now released, and is signed using shim.
Follow the instructions on https://ipxe.org/secboot for how to use this.
This is linked from the release page https://github.com/ipxe/ipxe/releases/tag/v2.0.0

**Reply by Falcon-Networks** [reply]
I had to disable SB to get mine to work with the default install. I am planning on adding third-party certs today to see if that makes any difference.

**Reply by NiKiZe** [reply]
Did you perhaps use/call ipxe.efi directly without using the shim?

**Reply by Falcon-Networks** [reply]
Yes sir, I am trying on a VM with SB enabled and EFI firmware on iPXE version 2.0.0

**Reply by NiKiZe** [reply]
For secure boot with default chain, you must boot shim, which is signed by Microsoft, that in turn loads, verifies, and runs ipxe.

**Reply by Falcon-Networks** [reply]
My apologies. To clarify, shimx64.efi fails with "Verification Failed: security violation". I figured you were talking about the x86_64-efi-SB binary ipxe.efi?
